ops/charm.py

```diff
@@ -305,28 +418,33 @@ def __init__(self, handle, relation, app=None, unit=None):
         self.app = app
         self.unit = unit
 
-    def snapshot(self) -> dict:
+    def snapshot(self) -> '_RelationEventSnapshot':
         """Used by the framework to serialize the event to disk.
 
         Not meant to be called by charm code.
         """
         snapshot = {
             'relation_name': self.relation.name,
             'relation_id': self.relation.id,
-        }
+        }  # type: 'RelationEvent._RelationEventSnapshot'
         if self.app:
             snapshot['app_name'] = self.app.name
         if self.unit:
             snapshot['unit_name'] = self.unit.name
         return snapshot
 
-    def restore(self, snapshot: dict) -> None:
+    def restore(self, snapshot: '_RelationEventSnapshot'):
         """Used by the framework to deserialize the event from disk.
 
         Not meant to be called by charm code.
         """
-        self.relation = self.framework.model.get_relation(
+        relation = self.framework.model.get_relation(
             snapshot['relation_name'], snapshot['relation_id'])
+        if relation is None:
+            raise TypeError(
+                'Unable to restore {}: relation {} (id={}) not found.'.format(
+                    self, snapshot['relation_name'], snapshot['relation_id']))
```

Comment:
This doesn't seem like it should be a type error.

Reply:
ValueError better? I often mix them up because I think 'the type of the thing is wrong --> typeError' ;/

Reply:
You mean, we should be allowing relation to be None?
Yeh, seems like it could be the case for relation-broken...
Not sure why I added it

Reply:
Probably because `.snapshot()` assumes that relation is not `None`.

Reply:
I think raising here is the right thing to do. The Snapshot / restore path will fail anyway otherwise, but with a less helpful trace.